File: agent/capabilities/remote_access.py
import platform
import logging
import subprocess
from typing import Dict, Any
from .base import BaseCapability

# winreg is a Windows-only stdlib module. Every call site below already
# checks platform.system() == "Windows" before touching it, but the bare
# import breaks importing this module (and the whole capabilities package)
# on Linux/macOS.
if platform.system() == "Windows":
    import winreg
else:
    winreg = None

logger = logging.getLogger(__name__)

class RemoteAccessCapability(BaseCapability):
    @property
    def capability_id(self) -> str:
        return "remote_access"

    @property
    def capability_name(self) -> str:
        return "Remote Access Control"

    def collect(self) -> Dict[str, Any]:
        return self.get_status()

    def get_description(self) -> str:
        return "Manage remote access settings (RDP)"

    def is_compatible(self, system_info: Dict[str, Any]) -> bool:
        # RDP is primarily for Windows in this context
        return system_info.get("os") == "Windows"

    def run(self, parameters: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Default run method - checks status.
        """
        return self.get_status()

    def get_status(self) -> Dict[str, Any]:
        if platform.system() != "Windows":
            return {"status": "unsupported", "enabled": False}

        try:
            # Check Registry for fDenyTSConnections (0 = Enabled, 1 = Disabled)
            key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"SYSTEM\CurrentControlSet\Control\Terminal Server", 0, winreg.KEY_READ)
            value, _ = winreg.QueryValueEx(key, "fDenyTSConnections")
            winreg.CloseKey(key)
            
            enabled = (value == 0)
            return {"status": "success", "enabled": enabled}
        except Exception as e:
            return {"status": "error", "error": str(e), "enabled": False}

    def enable_rdp(self) -> Dict[str, Any]:
        if platform.system() != "Windows":
            return {"status": "error", "error": "Only supported on Windows"}

        try:
            # 1. Enable in Registry
            key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"SYSTEM\CurrentControlSet\Control\Terminal Server", 0, winreg.KEY_SET_VALUE)
            winreg.SetValueEx(key, "fDenyTSConnections", 0, winreg.REG_DWORD, 0)
            winreg.CloseKey(key)

            # 2. Allow in Firewall (using netsh)
            # "netsh advfirewall firewall set rule group="remote desktop" new enable=Yes"
            cmd = ['netsh', 'advfirewall', 'firewall', 'set', 'rule', 'group=remote desktop', 'new', 'enable=Yes']
            logger.info(f"Executing: {' '.join(cmd)}")
            subprocess.run(cmd, capture_output=True, check=True)
            
            return {"status": "success", "message": "RDP Enabled successfully", "enabled": True}
        except Exception as e:
            logger.error(f"Failed to enable RDP: {e}")
            return {"status": "error", "error": str(e)}

    def disable_rdp(self) -> Dict[str, Any]:
        if platform.system() != "Windows":
            return {"status": "error", "error": "Only supported on Windows"}

        try:
            # 1. Disable in Registry
            key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"SYSTEM\CurrentControlSet\Control\Terminal Server", 0, winreg.KEY_SET_VALUE)
            winreg.SetValueEx(key, "fDenyTSConnections", 0, winreg.REG_DWORD, 1)
            winreg.CloseKey(key)
            
            return {"status": "success", "message": "RDP Disabled successfully", "enabled": False}
        except Exception as e:
            logger.error(f"Failed to disable RDP: {e}")
            return {"status": "error", "error": str(e)}

    def start_desktop_stream(self, session_id: str, url: str, tenant_key: str = "",
                             requester_name: str = "", requester_email: str = "",
                             tenant_name: str = "", control: bool = False):
        """
        Start sending desktop screenshots to the WebSocket URL.
        If `control` is True, run the interactive-control consent flow (D-01..D-12):
        ask the endpoint user to accept before any input is replayed, then
        replay gated browser→agent input frames and show the persistent stop bar.
        """
        import threading
        import websocket
        import json
        import time
        import base64
        import io
        import mss

        ws_headers = [f"X-Tenant-Key: {tenant_key}"] if tenant_key else None
        from PIL import Image

        def send_json(ws, obj):
            try:
                ws.send(json.dumps(obj))
            except Exception as e:
                logger.warning(f"send_json: {e}")

        def run_stream():
            logger.info(f"Starting desktop stream for session {session_id} to {url}")

            # Interactive-control consent flow (D-01): no SendInput before the
            # endpoint user accepts. Consent runs on a daemon thread so the
            # socket can keep processing control_stop / admin force-close while
            # the endpoint user decides. State messages are only sent once ws
            # exists (on open) — a pre-open send hits a reference-before-
            # assignment NameError that previously killed the whole stream.
            from . import consent_ui
            from .control_input import ConsentGate, parse_input_frame, replay_input_event

            gate = ConsentGate()
            stop_bar_shown = False

            def on_message(ws, message):
                # Only input frames from an accepted+active control session
                # are replayed, and only while the gate authorises (D-01).
                if not control or not gate.authorised():
                    return
                try:
                    if isinstance(message, bytes):
                        message = message.decode("utf-8")
                    ev = parse_input_frame(message)
                    replay_input_event(ev)
                except ValueError as e:
                    logger.warning(f"input frame rejected: {e}")

            def on_error(ws, error):
                logger.error(f"Desktop Stream WebSocket Error: {error}")

            def on_close(ws, close_status_code, close_msg):
                logger.info("Desktop Stream WebSocket Closed.")
                if stop_bar_shown:
                    try:
                        consent_ui.hide_stop_bar(session_id)
                    except Exception:
                        pass

            def consent_flow(ws):
                # Runs on a daemon thread; concedes the gate (or ends the
                # session) once the endpoint user decides.
                nonlocal stop_bar_shown
                if platform.system() != "Windows":
                    send_json(ws, {"type": "error", "reason": "unsupported_platform",
                                   "message": "interactive control requires Windows"})
                    return
                try:
                    decision = consent_ui.request_consent(
                        session_id, requester_name, requester_email,
                        tenant_name, timeout_secs=60)
                except consent_ui.ConsentError as ce:
                    reason = "no_interactive_desktop" if "no interactive" in ce.reason else (
                        "consent_declined")
                    send_json(ws, {"type": "error", "reason": reason, "message": ce.reason})
                    return
                if decision == consent_ui.ConsentDecision.Accept:
                    gate.concede()
                    send_json(ws, {"type": "control_state", "state": "active",
                                   "message": "consent granted"})
                    try:
                        consent_ui.show_stop_bar(session_id, requester_name)
                        stop_bar_shown = True
                    except consent_ui.ConsentError:
                        pass
                else:
                    reason = ("consent_timeout" if decision == consent_ui.ConsentDecision.Timeout
                             else "consent_declined")
                    send_json(ws, {"type": "control_state", "state": "ended",
                                   "reason": reason, "message": "consent not granted"})

            def on_open(ws):
                logger.info("Desktop Stream Connected. Starting frame capture.")
                if control:
                    send_json(ws, {"type": "control_state", "state": "awaiting_consent",
                                   "message": "consent required before control"})
                    threading.Thread(target=consent_flow, args=(ws,), daemon=True).start()

                with mss.mss() as sct:
                    # Use first monitor
                    monitor = sct.monitors[1]

                    while True:
                        # Endpoint stop bar one-click revocation (D-09).
                        if control and consent_ui.stop_requested(session_id):
                            gate.revoke()
                            send_json(ws, {"type": "control_state", "state": "ended",
                                           "reason": "user_stopped",
                                           "message": "control stopped by endpoint user"})
                            if stop_bar_shown:
                                try:
                                    consent_ui.hide_stop_bar(session_id)
                                except Exception:
                                    pass
                            break

                        # No screen before consent (D-01): a refused control
                        # session must not leak frames to the requester.
                        if control and not gate.authorised():
                            time.sleep(0.3)
                            continue

                        try:
                            # Capture
                            sct_img = sct.grab(monitor)

                            # Convert to PIL Image
                            img = Image.frombytes("RGB", sct_img.size, sct_img.bgra, "raw", "BGRX")

                            # Resize for performance (max 800px width)
                            max_width = 800
                            if img.width > max_width:
                                ratio = max_width / img.width
                                new_size = (max_width, int(img.height * ratio))
                                img = img.resize(new_size, Image.Resampling.LANCZOS)

                            # Compress to JPEG
                            buffer = io.BytesIO()
                            img.save(buffer, format="JPEG", quality=50, optimize=True)
                            b64_data = base64.b64encode(buffer.getvalue()).decode('utf-8')

                            # Send
                            payload = {
                                "type": "frame",
                                "timestamp": time.time(),
                                "data": b64_data
                            }
                            ws.send(json.dumps(payload))

                            # Limit FPS (~5-10 FPS)
                            time.sleep(0.15)

                        except Exception as capture_err:
                            logger.error(f"Frame Capture Error: {capture_err}")
                            time.sleep(1)  # Backoff

            # Connection
            ws = websocket.WebSocketApp(
                url,
                header=ws_headers,
                on_open=on_open,
                on_message=on_message,
                on_error=on_error,
                on_close=on_close
            )
            try:
                ws.run_forever()
            except Exception as e:
                logger.error(f"Failed to start desktop stream: {e}")

        thread = threading.Thread(target=run_stream, daemon=True)
        thread.start()

        return {"status": "success", "message": "Desktop stream thread started"}

    def start_reverse_shell(self, session_id: str, url: str, tenant_key: str = ""):
        """
        Start a reverse shell connected to the WebSocket URL.
        Spawns a new thread to handle the connection to avoid blocking the agent.
        """
        import threading
        import websocket
        import os

        ws_headers = [f"X-Tenant-Key: {tenant_key}"] if tenant_key else None

        def run_shell():
            logger.info(f"Starting reverse shell for session {session_id} to {url}")
            
            # Determine shell
            system = platform.system()
            if system == "Windows":
                shell_cmd = ["powershell.exe", "-NoLogo"]
            else:
                shell_cmd = ["/bin/bash", "-i"]
                
            try:
                # Start process unbuffered
                process = subprocess.Popen(
                    shell_cmd,
                    stdin=subprocess.PIPE,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT,
                    bufsize=0,
                    shell=False
                )
                
                def on_message(ws, message):
                    try:
                        if process.poll() is not None:
                            ws.close()
                            return
                        # Write to process stdin
                        if isinstance(message, str):
                            message = message.encode('utf-8')
                        process.stdin.write(message)
                        process.stdin.flush()
                    except Exception as e:
                        logger.error(f"Shell Input Error: {e}")

                def on_error(ws, error):
                    logger.error(f"WebSocket Error: {error}")

                def on_close(ws, close_status_code, close_msg):
                    logger.info("WebSocket Closed. Terminating shell.")
                    process.terminate()

                def on_open(ws):
                    logger.info("WebSocket Connection Opened. Starting output thread.")
                    
                    def forward_output():
                        try:
                            while True:
                                output = process.stdout.read(1024)
                                if not output:
                                    break
                                ws.send(output.decode('utf-8', errors='replace'))
                        except Exception as e:
                            logger.error(f"Shell Output Error: {e}")
                        finally:
                            ws.close()
                            
                    threading.Thread(target=forward_output, daemon=True).start()

                # Connection
                ws = websocket.WebSocketApp(
                    url,
                    header=ws_headers,
                    on_open=on_open,
                    on_message=on_message,
                    on_error=on_error,
                    on_close=on_close
                )
                ws.run_forever()
                
            except Exception as e:
                logger.error(f"Failed to start reverse shell: {e}")

        # Start the WebSocket client in a separate thread so it doesn't block the main Agent loop
        thread = threading.Thread(target=run_shell, daemon=True)
        thread.start()
        
        return {"status": "success", "message": "Reverse shell thread started"}

File: agent-install/omni-agent-rs/src/capabilities/remote_access.rs
use super::Capability;
use serde_json::{json, Value};
use sysinfo::System;

/// Identity of the admin requesting interactive control. Rendered in the consent
/// dialog and stop bar, and POSTed with the consent decision (T-74-19 audit).
#[derive(Default, Clone)]
pub struct Requester {
    pub name: String,
    pub email: String,
    pub tenant: String,
}

pub struct RemoteAccessCapability;

impl Capability for RemoteAccessCapability {
    fn id(&self) -> &'static str { "remote_access" }
    fn name(&self) -> &'static str { "Remote Access Control" }

    fn collect(&self, _sys: &System) -> Value {
        let status = rdp_status();
        json!({
            "rdp_enabled": status.0,
            "rdp_port": 3389,
            "status": if status.1.is_empty() { "success" } else { "error" },
            "error": status.1,
            "timestamp": chrono::Utc::now().to_rfc3339(),
        })
    }
}

fn rdp_status() -> (bool, String) {
    #[cfg(windows)]
    {
        use winreg::enums::*;
        use winreg::RegKey;
        match RegKey::predef(HKEY_LOCAL_MACHINE)
            .open_subkey(r"SYSTEM\CurrentControlSet\Control\Terminal Server")
            .and_then(|k| k.get_value::<u32, _>("fDenyTSConnections"))
        {
            Ok(val) => (val == 0, String::new()),
            Err(e) => (false, e.to_string()),
        }
    }
    #[cfg(not(windows))]
    {
        (false, "unsupported on this platform".to_string())
    }
}

pub fn set_rdp(enable: bool) -> Result<(), String> {
    #[cfg(windows)]
    {
        use winreg::enums::*;
        use winreg::RegKey;
        let key = RegKey::predef(HKEY_LOCAL_MACHINE)
            .open_subkey_with_flags(
                r"SYSTEM\CurrentControlSet\Control\Terminal Server",
                KEY_SET_VALUE,
            )
            .map_err(|e| e.to_string())?;
        let deny: u32 = if enable { 0 } else { 1 };
        key.set_value("fDenyTSConnections", &deny)
            .map_err(|e| e.to_string())?;
        let action = if enable { "Yes" } else { "No" };
        let _ = std::process::Command::new("netsh")
            .args([
                "advfirewall", "firewall", "set", "rule",
                "group=remote desktop", "new", &format!("enable={}", action),
            ])
            .output();
        log::info!("RDP {}", if enable { "enabled" } else { "disabled" });
        Ok(())
    }
    #[cfg(not(windows))]
    {
        Err("unsupported on this platform".to_string())
    }
}

/// Build a WebSocket client request carrying the tenant's `X-Tenant-Key`
/// header. `tunnel_agent_side` (backend/tunnel_endpoints.py) requires either
/// a valid JWT `?token=` or this header to authenticate the agent side of
/// the tunnel — the agent has never had a JWT of its own, and the previous
/// bare `connect_async(url)` sent neither, so every tunnel connection was
/// silently rejected with code 4401 before any shell/capture logic ran.
fn tunnel_request(
    url: &str,
    tenant_key: &str,
) -> Result<tokio_tungstenite::tungstenite::handshake::client::Request, Box<dyn std::error::Error + Send + Sync>> {
    use tokio_tungstenite::tungstenite::client::IntoClientRequest;
    let mut req = url.into_client_request()?;
    req.headers_mut().insert(
        "X-Tenant-Key",
        tokio_tungstenite::tungstenite::http::HeaderValue::from_str(tenant_key)?,
    );
    Ok(req)
}

/// Spawn a WebSocket-based reverse shell in a background tokio task.
pub fn start_reverse_shell(session_id: String, url: String, tenant_key: String) {
    tokio::spawn(async move {
        log::info!("Reverse shell starting: session={session_id} url={url}");
        if let Err(e) = reverse_shell_run(&url, &tenant_key).await {
            log::error!("Reverse shell error: {e}");
        }
        log::info!("Reverse shell ended: session={session_id}");
    });
}

/// Spawn a desktop streaming task that sends JPEG frames over WebSocket.
/// When `control` is true the same tunnel also relays browser input frames
/// back to the agent, gated behind endpoint-user consent (D-01).
pub fn start_desktop_stream(
    session_id: String,
    url: String,
    tenant_key: String,
    control: bool,
    requester: Requester,
) {
    tokio::spawn(async move {
        log::info!(
            "Desktop stream starting: session={session_id} url={url} control={control}"
        );
        if let Err(e) = desktop_stream_run(&session_id, &url, &tenant_key, control, &requester).await {
            log::error!("Desktop stream error: {e}");
        }
        log::info!("Desktop stream ended: session={session_id}");
    });
}

fn spawn_local_shell() -> std::io::Result<tokio::process::Child> {
    #[cfg(windows)]
    return tokio::process::Command::new("powershell.exe")
        .args(["-NoLogo", "-NonInteractive"])
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn();

    #[cfg(not(windows))]
    tokio::process::Command::new("/bin/bash")
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
}

async fn reverse_shell_run(url: &str, tenant_key: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    use futures_util::{SinkExt, StreamExt};
    use tokio::io::{AsyncReadExt, AsyncWriteExt};
    use tokio::sync::mpsc;
    use tokio_tungstenite::{connect_async, tungstenite::Message};

    let (ws_stream, _) = connect_async(tunnel_request(url, tenant_key)?).await?;
    let (ws_write, mut ws_read) = ws_stream.split();
    let mut child = spawn_local_shell()?;
    let mut proc_stdin = child.stdin.take().ok_or("no stdin")?;
    let proc_stdout = child.stdout.take().ok_or("no stdout")?;
    let (tx, mut rx) = mpsc::channel::<Message>(32);

    // Relay process stdout → WebSocket
    tokio::spawn(async move {
        let mut buf = [0u8; 4096];
        let mut reader = proc_stdout;
        loop {
            match reader.read(&mut buf).await {
                Ok(0) | Err(_) => break,
                Ok(n) => {
                    let text = String::from_utf8_lossy(&buf[..n]).to_string();
                    if tx.send(Message::Text(text.into())).await.is_err() { break; }
                }
            }
        }
    });

    tokio::spawn(async move {
        let mut sink = ws_write;
        while let Some(msg) = rx.recv().await {
            if sink.send(msg).await.is_err() { break; }
        }
    });

    // Relay WebSocket input → process stdin
    while let Some(Ok(msg)) = ws_read.next().await {
        match msg {
            Message::Text(text) => {
                if proc_stdin.write_all(text.as_bytes()).await.is_err() { break; }
            }
            Message::Binary(data) => {
                if proc_stdin.write_all(&data).await.is_err() { break; }
            }
            Message::Close(_) => break,
            _ => {}
        }
    }

    let _ = child.kill().await;
    Ok(())
}


// Windows: long-lived PowerShell process captures JPEG frames and emits base64 lines.
// Phase 74 consent-gate flow:
//   1. Connect tunnel, send control_state:"awaiting_consent"
//   2. spawn_blocking request_consent (blocks on endpoint user)
//   3. POST decision to backend audit endpoint (T-74-19)
//   4. On accept: show_stop_bar, send control_state:"active", start capture + input loop
//   5. Input loop: rate cap 50/s, ConsentGate.authorised() per frame, stop_requested() per frame
//   6. On decline/timeout/stop: send control_state:"ended", cleanup, exit
#[cfg(windows)]
async fn desktop_stream_run(
    session_id: &str,
    url: &str,
    tenant_key: &str,
    control: bool,
    requester: &Requester,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    use crate::capabilities::control_input::{control_input_replay, parse_input_frame, ConsentGate};
    use crate::consent_ui::{request_consent, show_stop_bar, hide_stop_bar, stop_requested, ConsentDecision, ConsentError};
    use futures_util::{SinkExt, StreamExt};
    use tokio::io::{AsyncBufReadExt, BufReader};
    use tokio::time::{timeout, Duration, interval};
    use tokio_tungstenite::{connect_async, tungstenite::Message};

    let (ws_stream, _) = connect_async(tunnel_request(url, tenant_key)?).await?;
    let (mut ws_write, ws_read) = ws_stream.split();

    // Build a control_state frame. `Message` is the WS frame wrapper; string
    // building avoids a lifetime-heavy closure returning a self-referential type.
    let send_state = |state: &str, reason: Option<&str>, message: Option<&str>| {
        let mut payload = format!(r#"{{"type":"control_state","state":"{}""#, state);
        if let Some(r) = reason { payload.push_str(&format!(r#","reason":"{}""#, r)); }
        if let Some(m) = message { payload.push_str(&format!(r#","message":"{}""#, m)); }
        payload.push('}');
        Message::Text(payload.into())
    };

    // 1. Signal awaiting consent
    let _ = ws_write.send(send_state("awaiting_consent", None, Some("Waiting for endpoint user consent"))).await;

    // 2. Request consent via spawn_blocking (blocks on endpoint UI)
    let session_id_owned = session_id.to_string();
    let requester_name = requester.name.clone();
    let requester_email = requester.email.clone();
    let tenant_name = requester.tenant.clone();
    let decision = tokio::task::spawn_blocking({
        let session_id = session_id_owned.clone();
        let name = requester_name.clone();
        let email = requester_email.clone();
        let tenant = tenant_name.clone();
        move || {
            request_consent(&session_id, &name, &email, &tenant, 60)
        }
    }).await.unwrap_or(Err(ConsentError::SpawnFailed("task join failed".into())));

    let decision = match decision {
        Ok(ConsentDecision::Accept) => ConsentDecision::Accept,
        Ok(ConsentDecision::Decline) => ConsentDecision::Decline,
        Ok(ConsentDecision::Timeout) => ConsentDecision::Timeout,
        Err(ConsentError::NoInteractiveSession) => {
            let _ = ws_write.send(Message::Text(r#"{"type":"error","reason":"no_interactive_desktop","message":"No interactive desktop session available (Session 0 isolation)"}"#.into())).await;
            return Err("no interactive desktop session".into());
        }
        Err(ConsentError::SpawnFailed(e)) => {
            let msg = e.replace('"', "'");
            let _ = ws_write.send(Message::Text(format!(r#"{{"type":"error","reason":"relay_failed","message":"{}"}}"#, msg).into())).await;
            return Err(e.into());
        }
    };

    // 3. POST decision to backend audit endpoint (T-74-19)
    let _ = post_consent_decision(url, &session_id_owned, &decision, requester).await;

    match decision {
        ConsentDecision::Decline => {
            let _ = ws_write.send(send_state("ended", Some("consent_declined"), Some("Endpoint user declined control"))).await;
            return Ok(());
        }
        ConsentDecision::Timeout => {
            let _ = ws_write.send(send_state("ended", Some("consent_timeout"), Some("Consent request timed out"))).await;
            return Ok(());
        }
        ConsentDecision::Accept => {}
    }

    // 4. Consent accepted — show stop bar, signal active
    if let Err(e) = show_stop_bar(&session_id_owned, &requester_name) {
        log::warn!("Failed to show stop bar: {e}");
    }
    let _ = ws_write.send(send_state("active", None, Some("Interactive control active"))).await;

    // 5. Start capture process
    let ps_script = r#"Add-Type -AssemblyName System.Windows.Forms,System.Drawing
$bounds = [System.Windows.Forms.Screen]::PrimaryScreen.Bounds
if ($bounds.Width -le 0 -or $bounds.Height -le 0) {
    [Console]::Out.WriteLine("ERR:No interactive desktop available (0x0 screen bounds) - the agent likely has no active user session to capture (Session 0 isolation)")
    [Console]::Out.Flush()
    exit 1
}
$enc = [System.Drawing.Imaging.ImageCodecInfo]::GetImageEncoders() | Where-Object {$_.MimeType -eq 'image/jpeg'} | Select-Object -First 1
$p = New-Object System.Drawing.Imaging.EncoderParameters 1
$p.Param[0] = New-Object System.Drawing.Imaging.EncoderParameter([System.Drawing.Imaging.Encoder]::Quality, [long]40)
while ($true) {
    try {
        $bm = New-Object System.Drawing.Bitmap([int]$bounds.Width, [int]$bounds.Height)
        $gr = [System.Drawing.Graphics]::FromImage($bm)
        $gr.CopyFromScreen($bounds.Location, [System.Drawing.Point]::Empty, $bounds.Size)
        $ms = New-Object System.IO.MemoryStream
        $bm.Save($ms, $enc, $p)
        $bm.Dispose(); $gr.Dispose()
        [Console]::Out.WriteLine([Convert]::ToBase64String($ms.ToArray()))
        [Console]::Out.Flush()
        $ms.Dispose()
    } catch {
        [Console]::Out.WriteLine("ERR:" + $_.Exception.Message)
        [Console]::Out.Flush()
    }
    Start-Sleep -Milliseconds 150
}"#;

    let mut child = match tokio::process::Command::new("powershell.exe")
        .args(["-NoProfile", "-NonInteractive", "-Command", ps_script])
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::null())
        .spawn()
    {
        Ok(child) => child,
        Err(e) => {
            let msg = format!("Failed to launch capture process: {}", e).replace('"', "'");
            let payload = format!(r#"{{"type":"error","message":"{}"}}"#, msg);
            let _ = ws_write.send(Message::Text(payload.into())).await;
            let _ = hide_stop_bar(&session_id_owned);
            return Err(format!("failed to spawn powershell.exe: {}", e).into());
        }
    };

    let stdout = match child.stdout.take() {
        Some(stdout) => stdout,
        None => {
            let _ = ws_write
                .send(Message::Text(
                    r#"{"type":"error","message":"Failed to capture stdout from the desktop capture process"}"#.into(),
                ))
                .await;
            let _ = child.kill().await;
            let _ = hide_stop_bar(&session_id_owned);
            return Err("no stdout from capture process".into());
        }
    };
    let mut lines = BufReader::new(stdout).lines();

    // First line must arrive within 5s
    let first_line = timeout(Duration::from_secs(5), lines.next_line()).await;
    let mut pending = match first_line {
        Ok(next) => next,
        Err(_) => {
            let _ = ws_write
                .send(Message::Text(
                    r#"{"type":"error","message":"No response from desktop capture process within 5s - it may lack an interactive desktop session"}"#.into(),
                ))
                .await;
            let _ = child.kill().await;
            let _ = hide_stop_bar(&session_id_owned);
            return Err("desktop capture timed out with no output".into());
        }
    };

    // 6. Authorisation gate, shared between the input loop and this loop so a
    //    revocation (stop sentinel / capture failure) is visible to the running,
    //    spawned input task. Concede ONCE after accept; every later SendInput is
    //    re-checked against the gate (D-01) before replay.
    let gate = std::sync::Arc::new(std::sync::Mutex::new(ConsentGate::new()));
    gate.lock().unwrap().concede(); // consent accepted

    let input_task = if control {
        let mut ws_read = ws_read;
        let gate_input = gate.clone();
        tokio::spawn(async move {
            let mut rate_limiter = interval(Duration::from_millis(20)); // 50 frames/sec cap
            while let Some(Ok(msg)) = ws_read.next().await {
                rate_limiter.tick().await;
                // Re-check authorisation before EVERY SendInput (D-01). The gate
                // is revoked by the capture loop on stop — a revoked gate refuses
                // further replay permanently within this session (D-12).
                if !gate_input.lock().unwrap().authorised() {
                    log::info!("Input gate revoked, ending input loop");
                    break;
                }
                let text = match msg {
                    Message::Text(t) => t.to_string(),
                    Message::Binary(b) => match String::from_utf8(b.to_vec()) {
                        Ok(t) => t,
                        Err(_) => {
                            log::warn!("control input: non-UTF8 binary frame dropped");
                            continue;
                        }
                    },
                    Message::Close(_) => break,
                    _ => continue,
                };
                match parse_input_frame(&text) {
                    Ok(ev) => control_input_replay(&ev),
                    Err(e) => log::warn!("control input: rejected frame: {e}"),
                }
            }
        })
    } else {
        tokio::spawn(async {})
    };

    // 7. Frame capture loop
    let mut end_reason = "unknown";
    let mut end_message = "Session ended";
    loop {
        // Check stop sentinel each frame (D-09/D-12): the endpoint user's Stop
        // Control click writes stop_{session}.stop. On fire, revoke the gate so
        // the concurrent input task stops replaying immediately.
        if stop_requested(&session_id_owned) {
            end_reason = "stop_requested";
            end_message = "Endpoint user stopped control";
            gate.lock().unwrap().revoke();
            break;
        }

        match pending {
            Ok(Some(ref line)) if line.starts_with("ERR:") => {
                let msg = line.trim_start_matches("ERR:").replace('"', "'");
                let payload = format!(r#"{{"type":"error","message":"{}"}}"#, msg);
                let _ = ws_write.send(Message::Text(payload.into())).await;
                end_reason = "capture_error";
                end_message = "Desktop capture failed";
                gate.lock().unwrap().revoke();
                break;
            }
            Ok(Some(ref line)) if !line.is_empty() => {
                let ts = chrono::Utc::now().timestamp_millis();
                let payload = format!(
                    r#"{{"type":"frame","timestamp":{},"data":"{}"}}"#,
                    ts, line
                );
                if ws_write.send(Message::Text(payload.into())).await.is_err() {
                    end_reason = "relay_failed";
                    end_message = "WebSocket relay failed";
                    gate.lock().unwrap().revoke();
                    break;
                }
            }
            Ok(None) | Err(_) => {
                end_reason = "capture_ended";
                end_message = "Desktop capture process ended";
                gate.lock().unwrap().revoke();
                break;
            }
            _ => {}
        }
        pending = lines.next_line().await;
    }

    // 8. Cleanup: abort input task, kill capture, hide stop bar, send ended state
    input_task.abort();
    let _ = child.kill().await;
    let _ = hide_stop_bar(&session_id_owned);
    let _ = ws_write.send(send_state("ended", Some(end_reason), Some(end_message))).await;

    Ok(())
}

/// POST consent decision to the backend audit endpoint (T-74-19).
/// Derives HTTP base from the tunnel WebSocket URL.
async fn post_consent_decision(
    tunnel_url: &str,
    session_id: &str,
    decision: &crate::consent_ui::ConsentDecision,
    requester: &Requester,
) {
    use reqwest::Client;
    // Derive HTTP endpoint from ws:// or wss:// tunnel URL
    let http_url = tunnel_url
        .replace("wss://", "https://")
        .replace("ws://", "http://")
        .replace("/tunnel", "/consent-decision")
        .replace("/viewer", "/consent-decision")
        .replace("/user", "/consent-decision");

    let body = serde_json::json!({
        "session_id": session_id,
        "decision": match decision {
            crate::consent_ui::ConsentDecision::Accept => "accept",
            crate::consent_ui::ConsentDecision::Decline => "decline",
            crate::consent_ui::ConsentDecision::Timeout => "timeout",
        },
        "requester_name": &requester.name,
        "requester_email": &requester.email,
        "tenant": &requester.tenant,
        "timestamp": chrono::Utc::now().to_rfc3339(),
    });

    let client = Client::new();
    if let Err(e) = client
        .post(&http_url)
        .json(&body)
        .timeout(std::time::Duration::from_secs(5))
        .send()
        .await
    {
        log::warn!("Failed to POST consent decision: {e}");
    }
}

#[cfg(not(windows))]
async fn desktop_stream_run(
    _session_id: &str,
    url: &str,
    tenant_key: &str,
    control: bool,
    _requester: &Requester,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    use futures_util::{SinkExt, StreamExt};
    use tokio_tungstenite::{connect_async, tungstenite::Message};

    // Open the tunnel just long enough to tell the viewer why no frames are
    // coming, instead of leaving it connected-but-silent forever (the
    // previous behavior: returning Err before ever calling connect_async
    // meant the viewer had zero signal that streaming would never start).
    let (ws_stream, _) = connect_async(tunnel_request(url, tenant_key)?).await?;
    let (mut ws_write, _) = ws_stream.split();
    // Control mode must say WHY it cannot proceed, so the browser can show a
    // real terminal state (T-74-04) rather than an endless spinner. View mode
    // keeps its original message untouched.
    let body = if control {
        r#"{"type":"error","reason":"unsupported_platform","message":"Interactive control is only supported on Windows agents"}"#
    } else {
        r#"{"type":"error","message":"Desktop streaming is only supported on Windows agents"}"#
    };
    let _ = ws_write.send(Message::Text(body.into())).await;
    Err("desktop streaming is only supported on Windows".into())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn post_consent_decision_url_derivation() {
        // HTTP base derives from the tunnel WebSocket URL (T-74-19)
        let http_url = "wss://relay.example.com/tunnel/abc"
            .replace("wss://", "https://")
            .replace("/tunnel", "/consent-decision");
        assert_eq!(
            http_url,
            "https://relay.example.com/consent-decision/abc"
        );
    }

    #[test]
    fn post_consent_decision_url_viewer_derivation() {
        let http_url = "ws://relay.example.com/viewer/xyz"
            .replace("ws://", "http://")
            .replace("/viewer", "/consent-decision")
            .replace("/user", "/consent-decision");
        assert_eq!(
            http_url,
            "http://relay.example.com/consent-decision/xyz"
        );
    }
}
